### PR TITLE
Raise exceptions in ohsome client instead of returning None in case of failed ohsome API query

### LINKED ISSUE #22: ohsome API definition should be independent from tailing /
It should be possible to set the ohsome API without the trailing `/`.

This is working:
`OHSOME_API=http://localhost:8082/`
This should work as well:
`OHSOME_API=http://localhost:8082`

### LINKED ISSUE #23: improve ohsome API client error handling
When sending false requests to the ohsome API, the errors are not handled properly. I think the ohsome client should throw an appropriate exception and not pass a HTTP error code 400 through like in this example: [test_indicator_tags_ratio.json](https://github.com/GIScience/ohsome-quality-analyst/blob/main/workers/tests/integrationtests/fixtures/vcr_cassettes/test_indicator_tags_ratio.json#L270-L315)

See also #24

### PR DESCRIPTION
**Pull request by matthiasschaub**
### Corresponding issue
Closes #23 
Closes #22 

### Checklist
- [x] I have updated my branch to `main` (e.g. through `git rebase main`)
- [ ] Rebase or squash commits in a reasonable manner
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)
- [x] @joker234 [Review `check_iso_time`](https://github.com/GIScience/ohsome-quality-analyst/pull/29/files#r638506388)
- [x] @joker234 [Review `test_ohsome_client.py`](https://github.com/GIScience/ohsome-quality-analyst/pull/29/files#diff-572732da8932515477f594a4e6234fdd3b65e2f703527d00fcec90db5bed4c69)
- [x] @matthiasschaub fix failing test https://jenkins.ohsome.org/job/ohsome-quality-analyst/job/PR-29/15/display/redirect

### CODE REVIEW

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 1]
```diff
@@ -20,52 +20,76 @@ async def query(
...
-    timestamp = datetime.datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%MZ")
-    return timestamp
+    # ohsome API response status codes are either 4xx and 5xx or 200
+    try:
+        return geojson.loads(resp.content)
+    except JSONDecodeError as error:
+        # ohsome API can return invalid GeoJSON after streaming of response
+        logging.error("Query ohsome API failed!")
+        raise error(
+            "Ohsome API returned invalid GeoJSON after streaming of the response"
```
Comment (matthiasschaub) [inline workers/ohsome_quality_analyst/ohsome/client.py:94]: Add reason: Timeout

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 2]
```diff
@@ -20,52 +20,76 @@ async def query(
...
     if time is not None:
         data["time"] = time
+    return data
 
-    logging.info("Query ohsome API.")
-    logging.debug("Query data: " + json.dumps(data, indent=4, sort_keys=True))
 
+async def _query_ohsome_api(url: str, data: dict):
     # set custom timeout as ohsome API takes a long time to send an answer
     timeout = httpx.Timeout(5, read=600)
```
Comment (matthiasschaub) [inline workers/ohsome_quality_analyst/ohsome/client.py:77]: Make timeout a bit longer than the timeout of ohsome API. Invalid response of ohsome API due to timeout is catched later in the code.

File: workers/tests/integrationtests/fixtures/vcr_cassettes/test_indicator_tags_ratio.json [hunk 3]
```diff
@@ -1,6 +1,226 @@
...
+                        "text/html; charset=iso-8859-1"
+                    ]
+                },
+                "content": "<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\n<html><head>\n<title>404 Not Found</title>\n</head><body>\n<h1>Not Found</h1>\n<p>The requested URL was not found on this server.</p>\n<hr>\n<address>Apache/2.4.29 (Ubuntu) Server at api.ohsome.org Port 443</address>\n</body></html>\n"
+            }
+        },
+        {
+            "request": {
+                "method": "POST",
+                "uri": "https://api.ohsome.org/v1elements/length/ratio",
```
Comment (joker234) [inline workers/tests/integrationtests/fixtures/vcr_cassettes/test_indicator_tags_ratio.json:172]: something went wrong there

Comment (matthiasschaub) [reply workers/tests/integrationtests/fixtures/vcr_cassettes/test_indicator_tags_ratio.json]: This was already fixed. But it seems like I forgot to recreate the VCR cassette.

**Review by joker234 — CHANGES_REQUESTED** [review]
Thanks for dealing with this issue. I left some suggestions, all in all a good change :+1:

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 4]
```diff
@@ -69,3 +62,36 @@ async def get_latest_ohsome_timestamp():
...
+    layer,
+    endpoint: Optional[str] = None,
+    ratio: bool = False,
+) -> str:
+    """Build endpoint URL of ohsome API."""
+    ohsome_api = OHSOME_API.removesuffix("/")
+    if endpoint is not None:
+        url = ohsome_api + "/" + endpoint.removesuffix("/")
+    else:
+        url = ohsome_api + "/" + layer.endpoint.removesuffix("/")
```
[the remark this answers is not in the document]
Comment (joker234) [reply workers/ohsome_quality_analyst/ohsome/client.py:77]: Ah, forgot that `removesuffix` was only introduced in Python 3.9 (compare [PEP 0616](https://www.python.org/dev/peps/pep-0616/)). I think we have to stick with `str.rstrip` for now.
```suggestion
    if endpoint is None:
        endpoint = layer.endpoint
    url = ohsome_api + "/" + endpoint.rstrip("/")
```

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 5]
```diff
@@ -69,3 +62,36 @@ async def get_latest_ohsome_timestamp():
...
     return timestamp
+
+
+def build_url(
+    layer,
+    endpoint: Optional[str] = None,
+    ratio: bool = False,
+) -> str:
+    """Build endpoint URL of ohsome API."""
+    ohsome_api = OHSOME_API.removesuffix("/")
```
Comment (joker234) [inline workers/ohsome_quality_analyst/ohsome/client.py:73]: See comment [below](https://github.com/GIScience/ohsome-quality-analyst/pull/29/files?file-filters%5B%5D=.md&file-filters%5B%5D=.py#r636146069).

```suggestion
    ohsome_api = OHSOME_API.rstrip("/")
```

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:77]: I was paying attention to when this function has been intrudced. Thanks for the suggestion!

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 6]
```diff
@@ -20,45 +20,38 @@ async def query(
...
+    logging.debug("Query filter: " + layer.filter)
+    logging.debug("Query data: " + json.dumps(data))
+    return await query_ohsome_api(url, data)
+
 
-    # set custom timeout as ohsome API takes a long time to send an answer
-    timeout = httpx.Timeout(5, read=600)
+async def query_ohsome_api(url: str, data: dict):
+    # set custom timeout as ohsome API takes a long time (10 minutes) to send an answer
+    timeout = httpx.Timeout(5, read=660)
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:34]: Because as far as I understand it if it is set to 10 minutes two cases can arise. Either httpx throughs the timeout error. Or Ohsome API chancels streaming of the response and we are left with an invalid JSON response. I wanted to make sure it is always the later case. Does this makes sense to you?

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 7]
```diff
@@ -20,45 +20,38 @@ async def query(
...
-            logging.warning("Ohsome query failed!")
-            return None
-    elif resp.status_code == 404:
-        logging.warning("Query failed!")
-        return None
+    try:
+        resp.raise_for_status()  # Raise for response status codes 4xx and 5xx
+    except httpx.HTTPStatusError:
+        logging.error("Query ohsome API failed!")
+        raise
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:42]: Can we provide more context then stating that quering the ohsome API failed since response status code could be any of 4xx and 5xx? Should we state more clearly that the reponse code has bee returend from the ohsome API?

File: workers/tests/unittests/test_ohsome_client.py [hunk 8]
```diff
@@ -40,15 +40,31 @@ def setUp(self) -> None:
 
     def test_query_valid_response(self) -> None:
         with patch("httpx.AsyncClient.post", new_callable=AsyncMock) as mock_request:
-            mock_request.return_value = httpx.Response(200, content=self.valid_geojson)
+            mock_request.return_value = httpx.Response(
+                200,
+                content=self.valid_geojson,
+                request=httpx.Request("POST", "mock.org"),
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:46]: Actually this will make a request to a website. It just create the object which is needed by the httpx client to make a request: https://www.python-httpx.org/api/#request

I substitude mock.org with "url" to make the generic naming more clear.

File: workers/tests/integrationtests/fixtures/vcr_cassettes/test_indicator_tags_ratio.json [hunk 9]
```diff
@@ -35,7 +35,7 @@
                 "http_version": "HTTP/1.1",
                 "headers": {
                     "Date": [
-                        "Fri, 14 May 2021 17:05:51 GMT"
+                        "Wed, 19 May 2021 15:14:22 GMT"
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/integrationtests/fixtures/vcr_cassettes/test_indicator_tags_ratio.json]: I removed this change from the PR.

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 10]
```diff
@@ -69,3 +62,36 @@ async def get_latest_ohsome_timestamp():
...
+    if endpoint is not None:
+        url = ohsome_api + "/" + endpoint.removesuffix("/")
+    else:
+        url = ohsome_api + "/" + layer.endpoint.removesuffix("/")
+    if ratio:
+        return url + "/" + "ratio"
+    return url
+
+
+def build_data_dict(
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:82]: I incorporated this suggestion into the code

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 11]
```diff
@@ -69,3 +62,36 @@ async def get_latest_ohsome_timestamp():
...
+    layer,
+    bpolys: str,
+    time: Optional[str] = None,
+    ratio: bool = False,
+) -> dict:
+    """Build data dictionary for ohsome API query."""
+    if ratio:
+        data = {"bpolys": bpolys, "filter": layer.filter, "filter2": layer.ratio_filter}
+    else:
+        data = {"bpolys": bpolys, "filter": layer.filter}
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:92]: I incorporated this suggestion into the code

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 12]
```diff
@@ -69,3 +62,36 @@ async def get_latest_ohsome_timestamp():
     timestamp_str = str(resp.json()["extractRegion"]["temporalExtent"]["toTimestamp"])
     timestamp = datetime.datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%MZ")
     return timestamp
+
+
+def build_url(
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:67]: I added unittests for this function.

[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:66]: Sorry I added tests for build_data_dir() but not yet build_url()

File: workers/tests/unittests/test_ohsome_client.py [hunk 13]
```diff
@@ -40,15 +40,31 @@ def setUp(self) -> None:
...
+
+    def test_query_status_code_400(self) -> None:
+        with patch("httpx.AsyncClient.post", new_callable=AsyncMock) as mock_request:
+            mock_request.return_value = httpx.Response(
+                400,
+                content=self.invalid_geojson,
+                request=httpx.Request("POST", "mock.org"),
+            )
+            with self.assertRaises(httpx.HTTPStatusError):
+                asyncio.run(ohsome_client.query(self.layer, self.bpolys))
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:71]: I am not sure if we want to tests the function like this because the value of the first part of the url (https://api.ohsome.org/v1 in this case) is depended on an evironment variable (OHSOME_API). But I guess we can simple read the env and use it.

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 14]
```diff
@@ -69,3 +61,74 @@ async def get_latest_ohsome_timestamp():
...
+                "Layer '{0}' has not 'ratio_filter' definied.".format(layer.name)
+            )
+        else:
+            data["filter2"] = layer.ratio_filter
+    if time is not None:
+        data["time"] = time
+    return data
+
+
+def check_iso_time(time: str) -> bool:
```
Comment (matthiasschaub) [inline workers/ohsome_quality_analyst/ohsome/client.py:102]: I started to implement this function to validate the time parameter before it is used as query parameter with the ohsome API. But since the ISO-8601 conform timestring(s) can be quite versatile it can only be tested rudimentary with the built-in stdlib. An alternative is to use a library such as dateutils to validate the timestring. Another one is to forgo it and pass it to the ohsome API in the hope to catch invalid timestring through the response of the ohsome API. 

@joker234 I would be happy to hear your thoughts on that.

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 15]
```diff
@@ -20,45 +20,38 @@ async def query(
...
-    logging.debug("Query data: " + json.dumps(data, indent=4, sort_keys=True))
+    logging.debug("Query URL: " + url)
+    logging.debug("Query filter: " + layer.filter)
+    logging.debug("Query data: " + json.dumps(data))
+    return await query_ohsome_api(url, data)
+
 
-    # set custom timeout as ohsome API takes a long time to send an answer
-    timeout = httpx.Timeout(5, read=600)
+async def query_ohsome_api(url: str, data: dict):
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:31]: I broke it down to make it easier to test.

Also when ohsomehex endpoints will be used we probably need to implement some kind of logic to decide which API will be queried. This could be done in the 'query' function.

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:65]: Now tests for build_url are present.

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py]: Does commit b390b848792f0df4aeb29d280e1ea197247ed32fb adresses the problem you raised here?

File: workers/tests/unittests/test_ohsome_client.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:72]: I took the ohsome api url as definied in definitions.py for the tests.

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 16]
```diff
@@ -69,3 +58,74 @@ async def get_latest_ohsome_timestamp():
...
+    bpolys: str,
+    time: Optional[str] = None,
+    ratio: bool = False,
+) -> dict:
+    """Build data dictionary for ohsome API query."""
+    data = {"bpolys": bpolys, "filter": layer.filter}
+    if ratio:
+        if layer.ratio_filter is None:
+            raise ValueError(
+                "Layer '{0}' has not 'ratio_filter' definied.".format(layer.name)
```
Comment (joker234) [inline workers/ohsome_quality_analyst/ohsome/client.py:90]:
```suggestion
                "Layer '{0}' has not 'ratio_filter' defined.".format(layer.name)
```

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 17]
```diff
@@ -8,10 +8,10 @@
 import httpx
 
 from ohsome_quality_analyst.utils.definitions import OHSOME_API
+from ohsome_quality_analyst.utils.exceptions import OhsomeApiError
 
 
-# TODO: Add documentation on time string format.
-# TODO: Add tests for ohsome package, including time = None leading to errors
+# TODO: Add more tests for ohsome package,
```
Comment (joker234) [inline workers/ohsome_quality_analyst/ohsome/client.py:14]:
```suggestion
# TODO: Add more tests for ohsome package.
```

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 18]
```diff
@@ -20,45 +20,34 @@ async def query(
...
-            logging.warning("Ohsome query failed!")
-            return None
-    elif resp.status_code == 404:
-        logging.warning("Query failed!")
-        return None
+    # ohsome API response status codes are either 4xx and 5xx or 200
+    try:
+        resp.raise_for_status()  # Raise for response status codes 4xx and 5xx
+    except httpx.HTTPStatusError as error:
+        raise OhsomeApiError("Ouery ohsome API failed!") from error
```
Comment (joker234) [inline workers/ohsome_quality_analyst/ohsome/client.py:42]:
```suggestion
        raise OhsomeApiError("Querying the ohsome API failed!") from error
```

**Review by joker234 — COMMENTED** [review]
Made some suggestions.
Additional ToDos (added to checklist):
 - Review check_iso_time
 - Review test_ohsome_client.py
 - fix failing test

File: workers/tests/unittests/test_ohsome_client.py [hunk 19]
```diff
@@ -40,15 +42,107 @@ def setUp(self) -> None:
 
     def test_query_valid_response(self) -> None:
         with patch("httpx.AsyncClient.post", new_callable=AsyncMock) as mock_request:
-            mock_request.return_value = httpx.Response(200, content=self.valid_geojson)
+            mock_request.return_value = httpx.Response(
+                200,
+                content=self.valid_geojson,
+                request=httpx.Request("POST", "url"),
```
Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:48]: I would suggest to use a valid URL here. I would even use the “correct” URL for your mocked result, e.g. `https://api.ohsome.org/v1/elements/geometry`.

File: workers/tests/unittests/test_ohsome_client.py [hunk 20]
```diff
@@ -40,15 +42,107 @@ def setUp(self) -> None:
...
+        self.assertTrue(ohsome_client.check_iso_time("2014-01-01"))
+        self.assertTrue(ohsome_client.check_iso_time("2007-01-25T12:00:00"))
+
+        self.assertTrue(
+            ohsome_client.check_iso_time("2014-01-01,2015-07-01,2018-10-10")
+        )
+
+        # TODO:
+        # self.assertTrue(ohsome_client.check_iso_time("2007-01-25T12:00:00Z"))
+        # self.assertTrue(ohsome_client.check_iso_time("2014-01-01/2018-01-01/P1Y"))
```
Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:88]: Before the PR is merged, I would suggest to either remove these two lines or make sure the either fail (`assertRaises`) or deliver the correct result.

Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:88]: Some edge cases are still missing which should be covered. E.g. time zones ;)

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 14 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:99]: I like the idea to catch invalid timestrings through the response of ohsome API.

While playing around with the ohsome API I found out two different response related to wrong use of the time parameter:

```
{
  "timestamp": "2021-05-31T07:33:31.206528",
  "status": 400,
  "message": "You need to give at least two timestamps or a time interval for this resource.",
  "requestUrl": "https://api.ohsome.org/v1/contributions/count?bboxes=8.67%2C49.39%2C8.71%2C49.42&filter=type%3Away%20and%20natural%3D*&format=json&time=20140101"
}

{
  "timestamp": "2021-05-31T07:36:40.661092",
  "status": 400,
  "message": "The provided time parameter is not ISO-8601 conform.",
  "requestUrl": "https://api.ohsome.org/v1/contributions/count?bboxes=8.67%2C49.39%2C8.71%2C49.42&filter=type%3Away%20and%20natural%3D*&format=json&time=20140101%2C2015010"
}
```

@tyrasd are there more cases that should be covered?

File: workers/tests/unittests/test_ohsome_client.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py]: Changed URL to http://www.example.org/

File: workers/tests/unittests/test_ohsome_client.py [hunk 21]
```diff
@@ -40,15 +40,31 @@ def setUp(self) -> None:
...
             self.assertTrue(response.is_valid)
 
     def test_query_invalid_response_with_status_code_200(self) -> None:
         """When response is streamed it can be invalid while status code equals 200"""
         with patch("httpx.AsyncClient.post", new_callable=AsyncMock) as mock_request:
             mock_request.return_value = httpx.Response(
-                200, content=self.invalid_geojson
+                200,
+                content=self.invalid_geojson,
+                request=httpx.Request("POST", "mock.org"),
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py]: Changed URL to http://www.example.org/

File: workers/tests/unittests/test_ohsome_client.py [hunk 22]
```diff
@@ -40,15 +40,31 @@ def setUp(self) -> None:
...
-            self.assertIsNone(response)
+            with self.assertRaises(JSONDecodeError):
+                asyncio.run(ohsome_client.query(self.layer, self.bpolys))
+
+    def test_query_status_code_400(self) -> None:
+        with patch("httpx.AsyncClient.post", new_callable=AsyncMock) as mock_request:
+            mock_request.return_value = httpx.Response(
+                400,
+                content=self.invalid_geojson,
+                request=httpx.Request("POST", "mock.org"),
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py]: Changed URL to http://www.example.org/

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 14 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py]: I removed functions for testing the time parameter. 
The code in line 43 will raise an exception if the ohsome response will return status code 400 when the time parameter is invalid.
```python
    try:
        resp.raise_for_status()  # Raise for response status codes 4xx and 5xx
    except httpx.HTTPStatusError as error:
        raise OhsomeApiError(
            "Querying the ohsome API failed!" + error.response.json()["message"]
        ) from error
```

Which will lead to following Error:

```
Traceback (most recent call last):
  File "/home/matthias/work/oqt/test.py", line 7, in <module>
    resp.raise_for_status()  # Raise for response status codes 4xx and 5xx
  File "/home/matthias/.cache/pypoetry/virtualenvs/ohsome-quality-analyst-JhxvPu0L-py3.9/lib/python3.9/site-packages/httpx/_models.py", line 1105, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: 400 Client Error:  for url: https://api.ohsome.org/v1/contributions/count?bboxes=8.67%2C49.39%2C8.71%2C49.42&filter=type%3Away%20and%20natural%3D*&format=json&time=20140101%2C2015010
For more information check: https://httpstatuses.com/400

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/matthias/work/oqt/test.py", line 10, in <module>
    raise OhsomeApiError("Querying the ohsome API failed! " + _) from error
ohsome_quality_analyst.utils.exceptions.OhsomeApiError: Querying the ohsome API failed! The provided time parameter is not ISO-8601 conform.
```

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 23]
```diff
@@ -69,3 +58,74 @@ async def get_latest_ohsome_timestamp():
...
+        timestrings = time.split(",")
+        for t in timestrings:
+            check_timestamp(t)
+
+    if not time:
+        logging.error("Could not parse timestring in ISO-8601 format")
+        raise ValueError
+    elif "," in time:
+        check_list_of_timestamps(time)
+    else:
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py]: This function has been removed in favor of raising an error when ohsome API reponse has status coe 400 (which is the case if the time parameter is invalid)

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 24]
```diff
@@ -69,3 +58,74 @@ async def get_latest_ohsome_timestamp():
...
+            logging.error("Could not parse timestring in ISO-8601 format")
+            raise
+
+    def check_list_of_timestamps(t):
+        timestrings = time.split(",")
+        for t in timestrings:
+            check_timestamp(t)
+
+    if not time:
+        logging.error("Could not parse timestring in ISO-8601 format")
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py]: This function has been removed in favor of raising an error when ohsome API reponse has status coe 400 (which is the case if the time parameter is invalid)

File: workers/tests/unittests/test_ohsome_client.py [hunk 25]
```diff
@@ -40,15 +42,107 @@ def setUp(self) -> None:
...
     def test_query_valid_response(self) -> None:
         with patch("httpx.AsyncClient.post", new_callable=AsyncMock) as mock_request:
-            mock_request.return_value = httpx.Response(200, content=self.valid_geojson)
+            mock_request.return_value = httpx.Response(
+                200,
+                content=self.valid_geojson,
+                request=httpx.Request("POST", "url"),
+            )
             response = asyncio.run(ohsome_client.query(self.layer, self.bpolys))
             self.assertTrue(response.is_valid)
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py]: It works and tests what should be tested, thats why I did not change it.

File: workers/tests/unittests/test_ohsome_client.py (hunk 20 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py]: This function has been removed in favor of raising an error when ohsome API reponse has status coe 400 (which is the case if the time parameter is invalid)

File: workers/tests/unittests/test_ohsome_client.py [hunk 26]
```diff
@@ -40,15 +42,107 @@ def setUp(self) -> None:
...
+        self.assertTrue(
+            ohsome_client.check_iso_time("2014-01-01,2015-07-01,2018-10-10")
+        )
+
+        # TODO:
+        # self.assertTrue(ohsome_client.check_iso_time("2007-01-25T12:00:00Z"))
+        # self.assertTrue(ohsome_client.check_iso_time("2014-01-01/2018-01-01/P1Y"))
+
+    def test_build_url(self) -> None:
+        ohsome_api = OHSOME_API.rstrip("/")
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py]: Yes, I see your point. I removed the dependecy on the env and rely on hardcoded URL for now.

**Review by joker234 — APPROVED** [review]
